index.src.html

```diff
@@ -2394,14 +2420,812 @@
+    15. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+    16. Resolve |promise| with |decodeResult|.
+
+: <dfn for=ImageDecoder>Resolve Decode</dfn> (with |promise| and |result|)
+:: 1. Queue a task on the [=control thread=] event loop to run these steps:
+        1. If {{ImageDecoder/[[closed]]}}, abort these steps.
+        2. Assert that |promise| is an element of
+            {{ImageDecoder/[[pending decode promises]]}}.
+        3. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        4. Resolve |promise| with |result|.
+
+: <dfn for=ImageDecoder>Reject Infeasible Decode</dfn> (with |promise|)
+:: 1. Assert that {{ImageDecoder/complete}} is `true` or
+        {{ImageDecoder/[[closed]]}} is `true`.
+    2. If {{ImageDecoder/complete}} is `true`, let |exception| be a
+            {{RangeError}}. Otherwise, let |exception| be an
+            {{InvalidStateError}} {{DOMException}}.
+    3. Queue a task on the [=control thread=] event loop to run these steps:
+        1. If {{ImageDecoder/[[closed]]}}, abort these steps.
+        2. Assert that |promise| is an element of
+            {{ImageDecoder/[[pending decode promises]]}}.
+        3. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        4. Reject |promise| with |exception|.
+
+: <dfn for=ImageDecoder>Reject Failed Decode</dfn> (with |promise|)
+:: 1. Queue a task on the [=control thread=] event loop to run these steps:
+        1. If {{ImageDecoder/[[closed]]}}, abort these steps.
+        2. Assert that |promise| is an element of
+            {{ImageDecoder/[[pending decode promises]]}}.
+        3. Remove |promise| from {{ImageDecoder/[[pending decode promises]]}}.
+        4. Reject |promise| with {{EncodingError}} {{DOMException}}.
+
+: <dfn for=ImageDecoder>Check Type Support</dfn> (with |type|)
+:: 1. If the user agent can provide a codec to support decoding |type|, return
+        `true`.
+    2. Otherwise, return `false`.
+
+: <dfn for=ImageDecoder>Reset ImageDecoder</dfn> (with |exception|)
+:: 1. Signal {{ImageDecoder/[[codec implementation]]}} to abort any active
+        decoding operation.
+    2. For each |metadataPromise| in
+        {{ImageDecoder/[[pending metadata promises]]}}:
+        1. Reject |metadataPromise| with |exception|.
+        2. Remove |metadataPromise| from
+            {{ImageDecoder/[[pending metadata promises]]}}.
+    3. For each |decodePromise| in
+        {{ImageDecoder/[[pending decode promises]]}}:
+        1. Reject |decodePromise| with |exception|.
+        2. Remove |decodePromise| from
+            {{ImageDecoder/[[pending decode promises]]}}.
+
+: <dfn for=ImageDecoder>Close ImageDecoder</dfn> (with |exception|)
+:: 1. Run the [=ImageDecoder/Reset ImageDecoder=] algorithm with |exception|.
+    1. Assign `true` to {{ImageDecoder/[[closed]]}}.
+    2. Clear {{ImageDecoder/[[codec implementation]]}} and release associated
```

Comment:
I should add an additional step here to close the ImageTrackList as well (needs internal slots in ImageTrack to indicate they've been detached in case any one still holds a ref).